File: examples/py/pogs/lasso.py
import h2ogpuml as h2ogpuml
from h2ogpuml.types import H2OFunctions
from numpy import abs, float32, float64, max, sqrt
from numpy.random import rand, randn

'''
Lasso

  minimize    (1/2) ||Ax - b||_2^2 + \lambda ||x||_1

See <h2ogpuml>/matlab/examples/lasso.m for detailed description.
template <typename T>
'''


def lasso(m, n, gpu=True, double_precision=False):
    # random matrix A
    A = randn(m, n)

    # cast A as float/double according to input args
    A = A if double_precision else float32(A)

    # true x vector, ~20% zeros
    x_true = (randn(n) / sqrt(n)) * float64(randn(n) < 0.8)

    # b= A*x_true + v (noise)
    b = A.dot(x_true) + 0.5 * randn(m)

    # lambda
    lambda_max = max(abs(A.T.dot(b)))

    # f(Ax) = ||Ax - b||_2^2
    f = h2ogpuml.FunctionVector(m, double_precision=double_precision)
    f.b[:] = b[:]
    f.h[:] = H2OFunctions.SQUARE

    # g(x) = 0.2*lambda_max*||x||_1
    g = h2ogpuml.FunctionVector(n, double_precision=double_precision)
    g.a[:] = 0.2 * lambda_max
    g.h[:] = H2OFunctions.ABS

    # use problem data A to create solver
    s = h2ogpuml.Pogs(A) if gpu else h2ogpuml.Pogs(A, n_gpus=0)

    # solve
    s.fit(f, g)

    # get solve time
    t = s.info.solvetime

    # tear down solver in C++/CUDA
    s.finish()

    return t


if __name__ == "__main__":
    print("Solve time:\t{:.2e} seconds".format(lasso(10000000, 100)))

File: examples/cpp/pogs/lasso.cpp
#include <stdio.h>
#include <random>
#include "matrix/matrix_dense.h"
#include "h2ogpumlglm.h"
#include "timer.h"

// Lasso
//   minimize    (1/2) ||Ax - b||_2^2 + \lambda ||x||_1
//
// See <h2ogpuml>/matlab/examples/lasso.m for detailed description.
template <typename T>
double Lasso(const std::vector<T>& A, const std::vector<T>& b) {
  size_t m=b.size();
  size_t n=A.size()/m;

  ////////////////////
  // set lambda for regularization
  ////////////////////
  T lambda_max = static_cast<T>(0);
#ifdef _OPENMP
#pragma omp parallel for reduction(max : lambda_max)
#endif
  for (unsigned int j = 0; j < n; ++j) {
    T u = 0;
    for (unsigned int i = 0; i < m; ++i)
      //u += A[i * n + j] * b[i];
      u += A[i + j * m] * b[i];
    lambda_max = std::max(lambda_max, std::abs(u));
  }

  ////////////////////
  // setup h2ogpuml
  ////////////////////
  fprintf(stderr,"MatrixDense\n"); fflush(stderr);
  h2ogpuml::MatrixDense<T> A_('r', m, n, A.data());
  fprintf(stderr,"H2OGPUMLDirect\n"); fflush(stderr);
  h2ogpuml::H2OGPUMLDirect<T, h2ogpuml::MatrixDense<T> > h2ogpuml_data(A_);
  fprintf(stderr,"f\n"); fflush(stderr);
  std::vector<FunctionObj<T> > f;
  fprintf(stderr,"g\n"); fflush(stderr);
  std::vector<FunctionObj<T> > g;

  f.reserve(m);
  for (unsigned int i = 0; i < m; ++i)
    f.emplace_back(kSquare, static_cast<T>(1), b[i]);

  g.reserve(n);
  for (unsigned int i = 0; i < n; ++i)
    g.emplace_back(kAbs, static_cast<T>(0.2) * lambda_max);

  double t = timer<double>();


  ////////////////////
  // Solve with h2ogpuml
  ////////////////////
  fprintf(stdout,"BEGIN SOLVE\n");
  if(0==1){ // debug
    //    h2ogpuml_data.SetAdaptiveRho(false); // trying
    //    h2ogpuml_data.SetRho(1.0);
    //  h2ogpuml_data.SetMaxIter(5u);
    //    h2ogpuml_data.SetMaxIter(1u);
    //    h2ogpuml_data.SetVerbose(4);
  }
  else if(1==1){ // debug
    //    h2ogpuml_data.SetAdaptiveRho(false); // trying
    //    h2ogpuml_data.SetEquil(false); // trying
    //    h2ogpuml_data.SetRho(1E-4);
    //    fprintf(stderr,"sets\n"); fflush(stderr);
    //    h2ogpuml_data.SetVerbose(4);
    //    h2ogpuml_data.SetMaxIter(5u);
  }
  else if(1==0){
    //    h2ogpuml_data.SetVerbose(4);
  }
#ifdef __CUDACC__
  //  cudaProfilerStart();
#endif
  fprintf(stderr,"Solve\n"); fflush(stderr);
  h2ogpuml_data.Solve(f, g);
#ifdef __CUDACC__
  //  cudaProfilerStop();
#endif
  double tf = timer<double>();
  fprintf(stdout,"END SOLVE: type 0 m %d n %d ts %g\n",(int)m,(int)n,tf-t);

  return tf-t;
}

template double Lasso<double>(const std::vector<double>& A, const std::vector<double>& b);
template double Lasso<float>(const std::vector<float>& A, const std::vector<float>& b);

